crates/ra_syntax/src/parsing/lexer.rs

```diff
@@ -14,91 +16,242 @@ pub struct Token {
+}
+
+/// Mapper function that converts `rustc_lexer::Token` with some additional context
+/// to `ParsedToken`
+fn rustc_token_to_parsed_token(
+    rustc_token: &rustc_lexer::Token,
+    text: &str,
+    token_start_offset: TextUnit,
+) -> ParsedToken {
+    // We drop some useful infromation here (see patterns with double dots `..`)
+    // Storing that info in `SyntaxKind` is not possible due to its layout requirements of
+    // being `u16` that come from `rowan::SyntaxKind` type and changes to `rowan::SyntaxKind`
+    // would mean hell of a rewrite
+
+    let token_range =
+        TextRange::offset_len(token_start_offset, TextUnit::from_usize(rustc_token.len));
+
+    let token_text = &text[token_range];
+
+    let (syntax_kind, error) = {
+        use rustc_lexer::TokenKind as TK;
+        use TokenizeError as TE;
+
+        match rustc_token.kind {
+            TK::LineComment => ok(COMMENT),
+            TK::BlockComment { terminated } => {
+                ok_if(terminated, COMMENT, TE::UnterminatedBlockComment)
+            }
+            TK::Whitespace => ok(WHITESPACE),
+            TK::Ident => ok(if token_text == "_" {
+                UNDERSCORE
+            } else {
+                SyntaxKind::from_keyword(token_text).unwrap_or(IDENT)
+            }),
+            TK::RawIdent => ok(IDENT),
+            TK::Literal { kind, .. } => match_literal_kind(&kind),
+            TK::Lifetime { starts_with_number } => {
+                ok_if(!starts_with_number, LIFETIME, TE::LifetimeStartsWithNumber)
+            }
+            TK::Semi => ok(SEMI),
+            TK::Comma => ok(COMMA),
+            TK::Dot => ok(DOT),
+            TK::OpenParen => ok(L_PAREN),
+            TK::CloseParen => ok(R_PAREN),
+            TK::OpenBrace => ok(L_CURLY),
+            TK::CloseBrace => ok(R_CURLY),
+            TK::OpenBracket => ok(L_BRACK),
+            TK::CloseBracket => ok(R_BRACK),
+            TK::At => ok(AT),
+            TK::Pound => ok(POUND),
+            TK::Tilde => ok(TILDE),
+            TK::Question => ok(QUESTION),
+            TK::Colon => ok(COLON),
+            TK::Dollar => ok(DOLLAR),
+            TK::Eq => ok(EQ),
+            TK::Not => ok(EXCL),
+            TK::Lt => ok(L_ANGLE),
+            TK::Gt => ok(R_ANGLE),
+            TK::Minus => ok(MINUS),
+            TK::And => ok(AMP),
+            TK::Or => ok(PIPE),
+            TK::Plus => ok(PLUS),
+            TK::Star => ok(STAR),
+            TK::Slash => ok(SLASH),
+            TK::Caret => ok(CARET),
+            TK::Percent => ok(PERCENT),
+            TK::Unknown => ok(ERROR),
+        }
     };
-    Some(Token { kind, len: TextUnit::from_usize(t.len) })
+
+    return ParsedToken {
+        token: Token { kind: syntax_kind, len: token_range.len() },
+        error: error
+            .map(|error| SyntaxError::new(SyntaxErrorKind::TokenizeError(error), token_range)),
+    };
+
+    type ParsedSyntaxKind = (SyntaxKind, Option<TokenizeError>);
```

Comment:
Defining the alias here probably takes up more space that just writing it out four times. Don't really care too much either way.

Reply:
Okay, I'll try to refactor it according to [this comment](https://github.com/rust-analyzer/rust-analyzer/pull/2911/files#r371165828)